vaadin-testbench-shared/src/main/java/com/vaadin/testbench/ElementQuery.java

```diff
@@ -400,6 +405,377 @@ public ElementQuery<T> withoutTheme(String theme) {
+        return withCondition(element -> (element instanceof HasPlaceholder hasPlaceholder) &&
+                comparison.test(hasPlaceholder.getPlaceholder(), text));
+    }
+
+    /**
+     * Requires the element's placeholder to exactly match the given placeholder value.
+     * <p>
+     * For partially matching text within the placeholder, see
+     * {@link #withPlaceholderContaining(String)}.
+     *
+     * @param placeholder
+     *            the placeholder to match
+     * @return this element query instance for chaining
+     *
+     * @see #withPlaceholderContaining(String)
+     * @see #withPlaceholder(String, BiPredicate)
+     */
+    public ElementQuery<T> withPlaceholder(String placeholder) {
+        return withPlaceholder(placeholder, String::equals);
+    }
+
+    /**
+     * Requires the element's placeholder to partially match the given text value.
+     * <p>
+     * For exactly matching the placeholder, see
+     * {@link #withPlaceholder(String)}.
+     *
+     * @param text
+     *          the text to match
+     * @return this element query instance for chaining
+     *
+     * @see #withPlaceholder(String)
+     * @see #withPlaceholder(String, BiPredicate)
+     */
+    public ElementQuery<T> withPlaceholderContaining(String text) {
+        return withPlaceholder(text, String::contains);
+    }
+
+    /**
+     * Requires the element's caption (i.e., label, placeholder, or text label)
+     * to satisfy the given comparison with the supplied text.
+     * <p>
+     * If the element has a label, the label value is used.
+     * If the element can have a label, but its value is empty and it has a placeholder,
+     * the placeholder value is used.
+     * If the element does not support labels nor placeholders
+     * but does support labels via its text (such as a button),
+     * the text is used.
+     * <p>
+     * For matching a caption exactly, see {@link #withCaption(String)},
+     * and for matching a caption partially, see {@link #withCaptionContaining(String)}.
+     * <p>
+     * This method can be used for other caption matching needs,
+     * such as performing a case-insensitive match:
+     * {@code withCaption("name", String::equalsIgnoreCase)}
+     * <br> a caption-ending match:
+     * {@code withCaption(" Name", String::endsWith)}
+     * <br> or a regular expression match:
+     * {@code withCaption("(First|Last) Name", String::matches)}
+     *
+     * @param text
+     *          the text to compare with the caption
+     * @param comparison
+     *          the comparison to use when comparing the caption
+     *          against the supplied text
+     *          (i.e., {@code comparison.test(elementCaption, text)}
+     * @return this element query instance for chaining
+     *
+     * @see #withCaption(String)
+     * @see #withCaptionContaining(String)
+     */
+    public ElementQuery<T> withCaption(String text, BiPredicate<String, String> comparison) {
```

Comment:
It looks like to me that this method behaves differently from others regarding handling of empty string for `text` parameter.
Here, if the element has an empty label/caption/text, the element is rejected without applying the `comparison`, whereas in other methods there's no such restriction.
For example, in `withPlaceholder` the `comparison` is applied without first checking if the element placeholder is empty.

We should always apply comparison, on the first non-empty value from the element, or empty if no attribute is valued.
Otherwise, we should not allow empty string for the `text` parameter, throwing an exception.
Probably the first one is more consistent with the rest of the other selectors.

Reply:
This difference in behavior is intentional. For `withLabel`, `withPlaceholder`, and `withText', we are explicitly looking at one and only one of those three places. However, `withCaption` is a convenience method—we don't know in which one of those three places the caption will be found, which is why when we find that particular value in the element to be empty, we move on to look at the next one.

I guess the one thing that makes less sense is that even though you should be able to match on an empty text value for `withLabel`, `withPlaceholder`, and `withText', what should that mean for captions? The way it works now is that if you provide an empty text value, it will incorrectly select an element with an empty label even if the element has a placeholder.

I think that is the thing I need to revise—in the special case when the input text value is empty, then all three places should be empty. I will make this adjustment. In addition, since none of the places can now ever have a null text value, I will disallow null text for each of them.